src/dinitctl.cc

```diff
@@ -1903,30 +1926,72 @@ static int enable_disable_service(int socknum, cpbuffer_t &rbuffer, service_dir_
     }
 
     if (socknum >= 0) {
+        if (verbose) {
+            cout << "Service '" << to << "' has been " << (enable ? "enabled" : "disabled") << "." << endl;
+        }
+
         // Check status of the service now
         auto m = membuf()
                 .append((char)cp_cmd::SERVICESTATUS)
                 .append(to_handle);
         write_all_x(socknum, m);
 
-        wait_for_reply(rbuffer, socknum);
+        int statussize = 6 + std::max(sizeof(pid_t), sizeof(int));;
+
+        // For an enable, we want to wait until the service has started so we can report any
+        // failure. But, if the service is already started, we won't get any service events, so we
+        // have to request status via SERVICESTATUS to catch that case. However, we may get
+        // a service event before the reply to SERVICESTATUS and in that case should use it to
+        // report status.
+        if (enable) {
+            int r = rbuffer.fill_to(socknum, 2);
+            while (r > 0 && rbuffer[0] >= 100) {
+                unsigned pktlen = (unsigned char) rbuffer[1];
+                fill_buffer_to(rbuffer, socknum, pktlen);
+
+                if (rbuffer[0] == (char)cp_info::SERVICEEVENT) {
+                    int ret = process_service_event(rbuffer, pktlen, to_handle, to, service_event_t::STARTED,
+                            service_event_t::STARTCANCELLED, false /* start */, verbose);
+                    if (ret >= 0) {
+                        // Make sure communication pipe is in a neutral state.
+                        wait_for_reply(rbuffer, socknum);
+                        if (rbuffer[0] != (char)cp_rply::SERVICESTATUS) {
+                            cerr << "dinitctl: protocol error." << endl;
+                            return 1;
+                        }
+                        fill_buffer_to(rbuffer, socknum, statussize + 1 /* reserved */);
+                        rbuffer.consume(2 + statussize);
```

Comment:
This consumes more than it potentially filled to, that can't be right. There should never be a call to `consume` with an amount that might be more than what is in the buffer, and, the call to `fill_buffer_to` needs to make sure that the entire packet has been read into the buffer.

It should be:

                        fill_buffer_to(rbuffer, socknum, 2 + statussize);
                        rbuffer.consume(2 + statussize);

I.e. same size in both cases, to read and consume a full packet.

Reply:
`fill_buffer_to` is documented:

    // Fill a circular buffer from a file descriptor, until it contains at least _rlength_ bytes.

That's _not_ in addition to the number of bytes already in the buffer - it's the total number of bytes. When `wait_for_reply` calls it with an argument of 1, it ensures there is at least 1 byte in the buffer. That doesn't affect later calls.

If you then call it with `statussize + 1` then there are least `statussize + 1` bytes in the buffer when it returns, and possibly _exactly_ that amount. That's not a full SERVICESTATUS packet, it's one byte short.